File: src/code/dart/merge_two_sorted_lists.dart
import 'structures/ListNode.dart';

class Solution {
  ListNode? mergeTwoLists(ListNode? list1, ListNode? list2) {
    if (list1 == null) return list2;
    if (list2 == null) return list1;

    ListNode? node1 = list1;
    ListNode? node2 = list2;
    ListNode resNode;

    if (node1.val <= node2.val) {
      resNode = node1;
      node1 = node1.next;
    } else {
      resNode = node2;
      node2 = node2.next;
    }

    ListNode? resList = resNode;

    while (node1 != null && node2 != null) {
      if (node1.val <= node2.val) {
        resNode.next = node1;
        resNode = node1;
        node1 = node1.next;
      } else {
        resNode.next = node2;
        resNode = node2;
        node2 = node2.next;
      }
    }

    if (node1 != null) resNode.next = node1;
    if (node2 != null) resNode.next = node2;

    return resList;
  }
}

main() {
  print(Solution().mergeTwoLists(
      ListNode(1, ListNode(2, ListNode(4))),
      ListNode(1, ListNode(3, ListNode(4)))));
  print(Solution().mergeTwoLists(
      ListNode(1, ListNode(2, ListNode(4))),
      ListNode(1, ListNode(8, ListNode(9)))));
  print(Solution().mergeTwoLists(null, null));
  print(Solution().mergeTwoLists(null, ListNode(0)));
}

File: src/code/kotlin/MergeTwoSortedLists.kt
package code.kotlin

import code.kotlin.structures.ListNode

class MergeTwoSortedLists {
    fun mergeTwoLists(list1: ListNode?, list2: ListNode?): ListNode? {
        if (list1 == null) return list2
        if (list2 == null) return list1

        if (list1.`val` < list2.`val`) {
            list1.next = mergeTwoLists(list1.next, list2)
            return list1
        } else {
            list2.next = mergeTwoLists(list1, list2.next)
            return list2
        }
    }
}

fun main() {
    val sol = MergeTwoSortedLists()::mergeTwoLists
    println(sol(ListNode(listOf(1, 2, 4)), ListNode(listOf(1, 3, 4))))
    println(sol(null, null))
}